includes/class-events-store.php

```diff
@@ -0,0 +1,592 @@
+<?php
+
+namespace Automattic\WP\Cron_Control;
+
+class Events_Store extends Singleton {
+	/**
+	 * PLUGIN SETUP
+	 */
+
+	/**
+	 * Class properties
+	 */
+	const TABLE_SUFFIX = 'a8c_cron_control_jobs';
+
+	const DB_VERSION        = 1;
+	const DB_VERSION_OPTION = 'a8c_cron_control_db_version';
+	const TABLE_CREATE_LOCK = 'a8c_cron_control_creating_table';
+
+	const STATUS_PENDING   = 'pending';
+	const STATUS_RUNNING   = 'running';
+	const STATUS_COMPLETED = 'complete';
+
+	const CACHE_KEY = 'a8c_cron_ctrl_option';
+
+	private $job_creation_suspended = false;
+
+	/**
+	 * Register hooks
+	 */
+	protected function class_init() {
+		// Check that the table exists and is the correct version
+		$this->prepare_tables();
+
+		// Option interception
+		add_filter( 'pre_option_cron', array( $this, 'get_option' ) );
+		add_filter( 'pre_update_option_cron', array( $this, 'update_option' ), 10, 2 );
+
+		// Disallow duplicates
+		add_filter( 'schedule_event', array( $this, 'block_creation_if_job_exists' ) );
+	}
+
+	/**
+	 * Build appropriate table name for this install
+	 */
+	public function get_table_name() {
+		global $wpdb;
+
+		return $wpdb->prefix . self::TABLE_SUFFIX;
+	}
+
+	/**
+	 * Build array of valid event statuses
+	 */
+	public function get_allowed_statuses() {
+		return array(
+			self::STATUS_PENDING,
+			self::STATUS_RUNNING,
+			self::STATUS_COMPLETED,
+		);
+	}
+
+	/**
+	 * Create the plugin's DB table when necessary
+	 */
+	protected function prepare_tables() {
+		// Nothing to do
+		if ( (int) get_option( self::DB_VERSION_OPTION ) === self::DB_VERSION ) {
+			return;
+		}
+
+		// Limit chance of race conditions when creating table
+		if ( false === wp_cache_add( self::TABLE_CREATE_LOCK, 1, null, 1 * \MINUTE_IN_SECONDS ) ) {
+			return;
+		}
+
+		// Use Core's method of creating/updating tables
+		if ( ! function_exists( 'dbDelta' ) ) {
+			require_once ABSPATH . '/wp-admin/includes/upgrade.php';
+		}
+
+		global $wpdb;
+
+		// Define schema and create the table
+		$schema = "CREATE TABLE `{$this->get_table_name()}` (
```

Comment:
I'd _really_ rather not. I'm familiar with the FG and related P2 discussion, and the arguments made there don't apply to this table, at least as I understand them. There's no use case for joining the plugin's table against any others--the IDs in `a8c_cron_control_jobs` don't correspond to any ID in any other table. If we found ourselves in a situation where a join was necessary, I'd rather we alter the table (and add appropriate indexes) than introduce a dependency on another table.

I'm also opposed to this change because I don't want to repeat the full column name throughout the codebase, nor do I really want to introduce a constant, helper, etc. to avoid repeating `a8c_cton_control_jobs_id`.

I understand the reasoning for tables on wpcom, or that might be joined against others, but don't feel it adds any value in this plugin.